File: windows/cockpit/P1Fast.Cockpit.Domain/Transport.cs
// Transport — escolhe entre cabo USB (primário) e Supabase Realtime (fallback)
// pra entregar IMU/GPS do iPhone ao cockpit do notebook Windows.
//
// Equivalente JS: web/cockpit/transport.js
// Spec: PLANO_FASE_1.md §6 MS-13.4 + ADR-023 amendment 2 (Flávio 2026-05-09)
//
// Princípio:
// - Primary  = cabo USB (TCP-over-USB via iproxy/usbmuxd) — latência 5-15 ms
// - Fallback = Supabase Realtime canal `live-stint-{stintId}`     — latência 150-500 ms
//
// Healthcheck por heartbeat 1 Hz; switch após silenceTimeoutMs (3 s);
// recovery com debounce de recoveryDebounceMs (1 s) antes de voltar pro primary.
//
// Modelo puro: zero IO. Os transportes concretos (LocalSocket, Realtime)
// implementam ITransport e são plugados no construtor.

namespace P1Fast.Cockpit.Domain;

public enum SelectorState
{
    OnPrimary,
    OnFallback,
    Recovering, // primary voltou a heartbeat, ainda em debounce
}

public enum TransportHealth
{
    Healthy,
    Silent,
    Starting,
}

public sealed record TransportEnvelope(double TMono, string Source, object? Payload);

/// <summary>Evento emitido por <c>onSwitch</c>. Stage é "state-change" | "recovery-started" | "transport-error".</summary>
public sealed record SwitchEvent(
    string Stage,
    SelectorState? From       = null,
    SelectorState? To         = null,
    string? Reason            = null,
    double? SincePrimary      = null,
    string? Which             = null,
    string? Err               = null
);

public interface ITransport
{
    string Name { get; }
    void Start(Action<TransportEnvelope> onMessage, Action<Exception> onError);
    void Stop();
}

/// <summary>Defaults da spec MS-2.8 + MS-13.4.</summary>
public static class TransportDefaults
{
    public const int HeartbeatExpectedMs  = 1000; // 1 Hz
    public const int SilenceTimeoutMs     = 3000; // 3 s sem heartbeat → switch
    public const int RecoveryDebounceMs   = 1000; // 1 s estável antes de voltar
}

public sealed record TransportSelectorStats(
    SelectorState State,
    TransportHealth PrimaryHealth,
    double? LastPrimaryHeartbeatAgoMs,
    int PacketsForwarded,
    int PacketsDropped,
    int SwitchesToPrimary,
    int SwitchesToFallback
);

public sealed class TransportSelector
{
    private readonly ITransport _primary;
    private readonly ITransport _fallback;
    private readonly Action<TransportEnvelope> _onMessage;
    private readonly Action<SwitchEvent>? _onSwitch;
    private readonly int _silenceTimeoutMs;
    private readonly int _recoveryDebounceMs;
    private readonly Func<double> _now;

    private SelectorState _state = SelectorState.OnPrimary;
    private TransportHealth _primaryHealth = TransportHealth.Starting;
    private double? _lastPrimaryHeartbeatAt;
    private double? _recoveryStartedAt;
    private bool _started;

    private int _packetsForwarded;
    private int _packetsDropped;
    private int _switchesToPrimary;
    private int _switchesToFallback;

    public TransportSelector(
        ITransport primary,
        ITransport fallback,
        Action<TransportEnvelope> onMessage,
        Action<SwitchEvent>? onSwitch = null,
        int silenceTimeoutMs   = TransportDefaults.SilenceTimeoutMs,
        int recoveryDebounceMs = TransportDefaults.RecoveryDebounceMs,
        Func<double>? now      = null)
    {
        ArgumentNullException.ThrowIfNull(primary);
        ArgumentNullException.ThrowIfNull(fallback);
        ArgumentNullException.ThrowIfNull(onMessage);

        _primary             = primary;
        _fallback            = fallback;
        _onMessage           = onMessage;
        _onSwitch            = onSwitch;
        _silenceTimeoutMs    = silenceTimeoutMs;
        _recoveryDebounceMs  = recoveryDebounceMs;
        _now                 = now ?? (() => (double)Environment.TickCount64);
    }

    public void Start()
    {
        if (_started) return;
        _started = true;
        _lastPrimaryHeartbeatAt = _now();
        _primary.Start(OnPrimaryMessage, e => OnTransportError("primary", e));
        _fallback.Start(OnFallbackMessage, e => OnTransportError("fallback", e));
    }

    public void Stop()
    {
        if (!_started) return;
        _started = false;
        try { _primary.Stop(); }  catch { /* swallow stop errors */ }
        try { _fallback.Stop(); } catch { /* swallow stop errors */ }
    }

    /// <summary>Avança o relógio interno; aplica regras de switch baseadas em silêncio do primary.</summary>
    public void Tick()
    {
        if (!_started) return;
        var now = _now();
        var sincePrimary = now - (_lastPrimaryHeartbeatAt ?? 0);

        if (_state == SelectorState.OnPrimary)
        {
            if (sincePrimary >= _silenceTimeoutMs)
                SwitchTo(SelectorState.OnFallback, reason: "silence", sincePrimary: sincePrimary);
            return;
        }

        if (_state == SelectorState.Recovering)
        {
            if (sincePrimary >= _silenceTimeoutMs)
            {
                _recoveryStartedAt = null;
                SwitchTo(SelectorState.OnFallback, reason: "recovery-aborted", sincePrimary: null);
                return;
            }
            var recoveryDuration = now - (_recoveryStartedAt ?? now);
            if (recoveryDuration >= _recoveryDebounceMs)
            {
                _recoveryStartedAt = null;
                SwitchTo(SelectorState.OnPrimary, reason: "recovered", sincePrimary: null);
            }
        }
    }

    public TransportSelectorStats GetStats()
    {
        double? ago = _lastPrimaryHeartbeatAt is null
            ? null
            : Math.Max(0, _now() - _lastPrimaryHeartbeatAt.Value);
        return new TransportSelectorStats(
            State:                       _state,
            PrimaryHealth:               _primaryHealth,
            LastPrimaryHeartbeatAgoMs:   ago,
            PacketsForwarded:            _packetsForwarded,
            PacketsDropped:              _packetsDropped,
            SwitchesToPrimary:           _switchesToPrimary,
            SwitchesToFallback:          _switchesToFallback);
    }

    // ── Handlers internos ──────────────────────────────────────────

    private void OnPrimaryMessage(TransportEnvelope env)
    {
        if (!_started) return;
        if (IsHeartbeat(env))
        {
            _lastPrimaryHeartbeatAt = _now();
            _primaryHealth = TransportHealth.Healthy;
            if (_state == SelectorState.OnFallback)
            {
                _state = SelectorState.Recovering;
                _recoveryStartedAt = _now();
                _onSwitch?.Invoke(new SwitchEvent(Stage: "recovery-started"));
            }
            return; // heartbeat não é forwarded
        }
        if (_state == SelectorState.OnPrimary || _state == SelectorState.Recovering)
        {
            _packetsForwarded++;
            _onMessage(env);
        }
        else
        {
            _packetsDropped++;
        }
    }

    private void OnFallbackMessage(TransportEnvelope env)
    {
        if (!_started) return;
        if (IsHeartbeat(env)) return; // heartbeat do fallback é só sinal de vida
        if (_state == SelectorState.OnFallback)
        {
            _packetsForwarded++;
            _onMessage(env);
        }
        else
        {
            _packetsDropped++;
        }
    }

    private void OnTransportError(string which, Exception err)
    {
        _onSwitch?.Invoke(new SwitchEvent(
            Stage: "transport-error",
            Which: which,
            Err:   err.Message));
    }

    private static bool IsHeartbeat(TransportEnvelope env) => env.Source == "heartbeat";

    private void SwitchTo(SelectorState newState, string reason, double? sincePrimary)
    {
        var prev = _state;
        if (prev == newState) return;
        _state = newState;
        if (newState == SelectorState.OnPrimary)  _switchesToPrimary++;
        if (newState == SelectorState.OnFallback) _switchesToFallback++;
        _onSwitch?.Invoke(new SwitchEvent(
            Stage:        "state-change",
            From:         prev,
            To:           newState,
            Reason:       reason,
            SincePrimary: sincePrimary));
    }
}

/// <summary>Transporte de mock pra teste/dev (sem IO real).</summary>
public sealed class MockTransport : ITransport
{
    public string Name { get; }

    private Action<TransportEnvelope>? _onMessage;
    private Action<Exception>? _onError;
    private bool _started;

    public MockTransport(string name)
    {
        Name = name;
    }

    public void Start(Action<TransportEnvelope> onMessage, Action<Exception> onError)
    {
        _onMessage = onMessage;
        _onError   = onError;
        _started   = true;
    }

    public void Stop() => _started = false;

    /// <summary>Helper de teste: simula pacote chegando do transporte.</summary>
    public void Inject(TransportEnvelope envelope)
    {
        if (!_started || _onMessage is null) return;
        _onMessage(envelope);
    }

    public void InjectError(Exception err) => _onError?.Invoke(err);
}

File: web/cockpit/transport.js
// Transport — escolhe entre cabo USB (primário) e Supabase Realtime (fallback)
// pra entregar IMU/GPS do iPhone ao cockpit web do notebook Windows.
//
// Spec: PLANO_FASE_1.md §6 MS-13.4 (reescrito 2026-05-09 amendment 2 — ADR-023)
// Mockup canônico: zero impacto (transport vive antes do CockpitState).
//
// Princípio:
// - "primary"  = cabo USB (TCP-over-USB via iproxy/usbmuxd) — latência 5-15 ms
// - "fallback" = Supabase Realtime canal `live-stint-{stintId}`     — latência 150-500 ms
//
// Healthcheck por heartbeat 1 Hz:
// - iPhone publica heartbeat `{ tMono, source: 'iphone-imu' }` em ambos canais
// - Selector marca primary HEALTHY ao receber heartbeat
// - Após `silenceTimeoutMs` (3 s) sem heartbeat do primary → switch pro fallback
// - Quando primary volta a heartbeatar, espera `recoveryDebounceMs` (1 s) com
//   heartbeats consistentes antes de voltar pro primary (evita flapping)
//
// Este módulo é PURO (zero IO de rede). Os transportes concretos
// (LocalSocketTransport, RealtimeTransport) implementam a interface abaixo
// e são plugados no construtor — modelo testável em Node.

export const TransportName = Object.freeze({
  PRIMARY:  'primary',
  FALLBACK: 'fallback',
});

export const TransportHealth = Object.freeze({
  HEALTHY:    'healthy',
  SILENT:     'silent',
  STARTING:   'starting',
});

export const SelectorState = Object.freeze({
  ON_PRIMARY:  'on-primary',
  ON_FALLBACK: 'on-fallback',
  RECOVERING:  'recovering', // primary voltou a heartbeat, ainda em debounce
});

// Defaults (PLANO §6 MS-2.8 + MS-13.4)
export const DEFAULTS = Object.freeze({
  heartbeatExpectedMs:  1000, // 1 Hz
  silenceTimeoutMs:     3000, // 3 s sem heartbeat → switch
  recoveryDebounceMs:   1000, // 1 s estável antes de voltar
});

/**
 * @typedef {Object} TransportEnvelope
 * @property {number} tMono              performance.now() do publisher
 * @property {string} source             'iphone-imu' | 't4000-can' | 'heartbeat' | ...
 * @property {Object} payload            o sample propriamente dito (IMU/GPS/T4000)
 */

/**
 * Interface esperada de um transporte concreto:
 *
 *   transport.start(onMessage, onError) → void
 *   transport.stop() → void
 *   transport.name (string)
 *
 * onMessage(envelope) é chamado a cada pacote recebido (incluindo heartbeat).
 * Implementações concretas: LocalSocketTransport, RealtimeTransport.
 */

// ── TransportSelector ──────────────────────────────────────────

export class TransportSelector {
  /**
   * @param {Object} opts
   * @param {Object} opts.primary    transporte primário (cabo)
   * @param {Object} opts.fallback   transporte fallback (Realtime)
   * @param {(env: any) => void} opts.onMessage   callback unificado
   * @param {(stage: string, info: any) => void} [opts.onSwitch]  observabilidade
   * @param {number} [opts.silenceTimeoutMs]
   * @param {number} [opts.recoveryDebounceMs]
   * @param {() => number} [opts.now]   injetável pra teste (ms monotônico)
   */
  constructor({
    primary,
    fallback,
    onMessage,
    onSwitch = () => {},
    silenceTimeoutMs = DEFAULTS.silenceTimeoutMs,
    recoveryDebounceMs = DEFAULTS.recoveryDebounceMs,
    now = () => (typeof performance !== 'undefined' ? performance.now() : Date.now()),
  }) {
    if (!primary || !fallback) throw new Error('TransportSelector: primary e fallback são obrigatórios');
    if (typeof onMessage !== 'function') throw new Error('TransportSelector: onMessage precisa ser função');

    this._primary = primary;
    this._fallback = fallback;
    this._onMessage = onMessage;
    this._onSwitch = onSwitch;
    this._silenceTimeoutMs = silenceTimeoutMs;
    this._recoveryDebounceMs = recoveryDebounceMs;
    this._now = now;

    this._state = SelectorState.ON_PRIMARY;
    this._primaryHealth = TransportHealth.STARTING;
    this._lastPrimaryHeartbeatAt = null;
    this._recoveryStartedAt = null;
    this._started = false;

    this._packetsForwarded = 0;
    this._packetsDropped = 0;
    this._switchesToPrimary = 0;
    this._switchesToFallback = 0;
  }

  start() {
    if (this._started) return;
    this._started = true;
    this._lastPrimaryHeartbeatAt = this._now();
    this._primary.start(
      (env) => this._onPrimaryMessage(env),
      (err) => this._onTransportError('primary', err),
    );
    this._fallback.start(
      (env) => this._onFallbackMessage(env),
      (err) => this._onTransportError('fallback', err),
    );
  }

  stop() {
    if (!this._started) return;
    this._started = false;
    try { this._primary.stop(); } catch {}
    try { this._fallback.stop(); } catch {}
  }

  /**
   * Avança o relógio interno (chamado por timer do app, ou pelo smoke pra forçar tick).
   * Aplica regras de switch baseadas em silêncio do primary.
   */
  tick() {
    if (!this._started) return;
    const now = this._now();
    const sincePrimary = now - (this._lastPrimaryHeartbeatAt ?? 0);

    // Se está no primary e ficou quieto demais → switch pro fallback
    if (this._state === SelectorState.ON_PRIMARY) {
      if (sincePrimary >= this._silenceTimeoutMs) {
        this._switchTo(SelectorState.ON_FALLBACK, { reason: 'silence', sincePrimary });
      }
      return;
    }

    // Se está em recovery e o primary continua heartbeatando estável → volta
    if (this._state === SelectorState.RECOVERING) {
      // Se primary silenciou no meio da recovery, aborta e volta pra fallback
      if (sincePrimary >= this._silenceTimeoutMs) {
        this._recoveryStartedAt = null;
        this._switchTo(SelectorState.ON_FALLBACK, { reason: 'recovery-aborted' });
        return;
      }
      const recoveryDuration = now - (this._recoveryStartedAt ?? now);
      if (recoveryDuration >= this._recoveryDebounceMs) {
        this._recoveryStartedAt = null;
        this._switchTo(SelectorState.ON_PRIMARY, { reason: 'recovered' });
      }
    }
  }

  getStats() {
    return {
      state: this._state,
      primaryHealth: this._primaryHealth,
      lastPrimaryHeartbeatAgoMs: this._lastPrimaryHeartbeatAt === null
        ? null : Math.max(0, this._now() - this._lastPrimaryHeartbeatAt),
      packetsForwarded: this._packetsForwarded,
      packetsDropped: this._packetsDropped,
      switchesToPrimary: this._switchesToPrimary,
      switchesToFallback: this._switchesToFallback,
    };
  }

  // ── handlers internos ────────────────────────────────────

  _onPrimaryMessage(env) {
    if (!this._started) return;
    if (this._isHeartbeat(env)) {
      this._lastPrimaryHeartbeatAt = this._now();
      this._primaryHealth = TransportHealth.HEALTHY;
      // se está no fallback e o primary heartbeatou, abre janela de recovery
      if (this._state === SelectorState.ON_FALLBACK) {
        this._state = SelectorState.RECOVERING;
        this._recoveryStartedAt = this._now();
        this._onSwitch('recovery-started', {});
      }
      return; // heartbeat não é forwarded
    }
    // payload normal: só forwarda se estado atual é primary OU recovering
    if (this._state === SelectorState.ON_PRIMARY || this._state === SelectorState.RECOVERING) {
      this._packetsForwarded++;
      this._onMessage(env);
    } else {
      this._packetsDropped++; // dado do primary chegando enquanto está no fallback (ignora pra evitar duplicata)
    }
  }

  _onFallbackMessage(env) {
    if (!this._started) return;
    if (this._isHeartbeat(env)) {
      // heartbeat do fallback é só sinal de vida — não influencia selector
      return;
    }
    // só forwarda fallback se realmente tá nele
    if (this._state === SelectorState.ON_FALLBACK) {
      this._packetsForwarded++;
      this._onMessage(env);
    } else {
      this._packetsDropped++;
    }
  }

  _onTransportError(which, err) {
    this._onSwitch('transport-error', { which, err: String(err && err.message || err) });
  }

  _isHeartbeat(env) {
    return env && env.source === 'heartbeat';
  }

  _switchTo(newState, info) {
    const prev = this._state;
    if (prev === newState) return;
    this._state = newState;
    if (newState === SelectorState.ON_PRIMARY) this._switchesToPrimary++;
    if (newState === SelectorState.ON_FALLBACK) this._switchesToFallback++;
    this._onSwitch('state-change', { from: prev, to: newState, ...info });
  }
}

// ── Transporte de mock pra teste/dev (sem IO real) ──────────

export class MockTransport {
  constructor(name) {
    this.name = name;
    this._onMessage = null;
    this._onError = null;
    this._started = false;
  }
  start(onMessage, onError) {
    this._onMessage = onMessage;
    this._onError = onError;
    this._started = true;
  }
  stop() { this._started = false; }
  /** Helper de teste: simula um pacote chegando do transporte. */
  inject(envelope) {
    if (!this._started || !this._onMessage) return;
    this._onMessage(envelope);
  }
  injectError(err) {
    if (this._onError) this._onError(err);
  }
}
